### PR TITLE
Inject mock services based on environment

### PR DESCRIPTION
**Pull request by radu-matei**
cc @seanknox, @flynnduism

### CODE REVIEW

File: src/app/project/project.component.ts [hunk 1]
```diff
@@ -1,13 +1,13 @@
-import { Component, OnInit, HostBinding } from '@angular/core';
+import { Component, OnInit, HostBinding, Inject } from '@angular/core';
```
Comment (seanknox) [inline src/app/project/project.component.ts]: What does `Inject` provide here?

File: src/app/services/build/MockBuildService.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { Build } from "../../models/project";
+import { Observable } from "rxjs/Observable";
+import { BuildService } from "./BuildService";
+
+export class MockBuildService implements BuildService {
+    getBuilds(projectId: string): Observable<Build[]> {
+        throw new Error("Method not implemented.");
+    }
```
Comment (seanknox) [inline src/app/services/build/MockBuildService.ts]: Is this where you'd imagine using `in-memory-web-api` to return the mock data? or something different?

File: src/environments/environment.ts [hunk 3]
```diff
@@ -1,8 +1,14 @@
...
 // The build system defaults to the dev environment which uses `environment.ts`, but if you do
 // `ng build --env=prod` then `environment.prod.ts` will be used instead.
 // The list of which env maps to which file can be found in `.angular-cli.json`.
 
 export const environment = {
-  production: false
+  production: false,
+
+  buildServiceType: MockBuildService,
+  projectServiceType: MockProjectService
```
Comment (seanknox) [inline src/environments/environment.ts]: I like this a lot. Very cool.

File: src/app/project/project.component.ts (hunk 1 above)
Comment (radu-matei) [reply src/app/project/project.component.ts]: Yeah, forgot to remove it after some tries...

Really miss `goimports` here 😄

File: src/app/services/build/MockBuildService.ts (hunk 2 above)
Comment (radu-matei) [reply src/app/services/build/MockBuildService.ts]: Either that, or simply strong type the existing mock data available in the `mock` directory.

**Comment by flynnduism** [comment]
Lovely work @radu-matei, this looks fab.

File: src/app/services/build/MockBuildService.ts (hunk 2 above)
Comment (seanknox) [reply src/app/services/build/MockBuildService.ts]: Do you recommend defining types as classes or interfaces? e.g. given this JSON:

```json
  {
    "id": "brigade-29d38c7477ecee18e184b69bec354fc350605c51bc16d4dd2b6073",
    "name": "technosophos/coffeesnob",
    "repo": {
      "name": "github.com/technosophos/coffeesnob",
      "cloneURL": "https://github.com/technosophos/coffeesnob.git"
    },
    "defaultScript": "",
    "kubernetes": {
      "namespace": "default",
      "vcsSidecar": "",
      "buildStorageSize": "50Mi"
    },
    "github": {
      "baseURL": "",
      "uploadURL": ""
    },
    "secrets": {},
    "worker": {
      "registry": "",
      "name": "",
      "tag": "",
      "pullPolicy": ""
    }
  }
```

```typescript
export class Project {
  constructor(
    public id: string,
    public name: string,
    public repo: Repo,
    public kubernetes: Kubernetes,
    public defaultScript: string,
    public github: any,
    public secrets: any,
    public worker: Worker
  ) {}
}
```

vs

```typescript
    export interface Project {
        id: string;
        name: string;
        repo: Repo;
        defaultScript: string;
        kubernetes: Kubernetes;
        defaultScript: string;
        github: Github;
        secrets: Secrets;
        worker: Worker;
    }
```

**Comment by seanknox** [comment]
@radu-matei can I have push rights to your branch?

**Comment by radu-matei** [comment]
@seanknox - added you and @flynnduism as collaborators on the repo, you should be able to push now.

File: src/app/project/project.component.html [hunk 4]
```diff
@@ -58,14 +58,16 @@ <h1>{{ project.name }} </h1>
...
             </span>
-            <span class="act-time-ago ng-binding">
-              {{ build.worker.status }}
-              <span>{{build.worker.end_time | amTimeAgo}}</span>.
-            </span>
-            <span class="act-time-duration">
-              Ran for
-              <span class="ng-binding">{{ build.worker.start_time | amDifference: build.worker.end_time :'seconds' : true | amDuration:'seconds' }}.</span>
-            </span>
+            <div *ngIf="build.worker">
```
Comment (seanknox) [inline src/app/project/project.component.html]: The API omits properties if they're empty, meaning we have to check for presence of properties throughout our code, which isn't great. I opened an issue on Brigade here about it: https://github.com/Azure/brigade/issues/385

**Comment by seanknox** [comment]
@flynnduism @radu-matei I think this is close to done. would love your thoughts.

- [x] can start app in dev or prod mode. prod will hit https://cors-anywhere.herokuapp.com/http://acid-api.technosophos.me:7745/v1 while dev will return mock data based on production. 
- [x] mock data is now typed
- [x] mock and real services return Observables
- [x] dashboard and project page functionality at parity with main `angular5` branch
- [ ] unit tests for services (and models? not sure if worth it)
- [ ] fix "_co.project is undefined" error on project page (this could be handled in another bug story)

kashti

File: src/app/models/Build.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+export interface Build {
+  id: string;
+  project_id: string;
+  type: string;
+  provider: string;
+  revision: Revision;
+  payload?: string;
+  script?: string;
+  worker?: BuildWorker;
```
Comment (seanknox) [inline src/app/models/Build.ts]: As an aside these properties were "optional" in that they're not always returned by the API had me running in circles for awhile.

**Pull request merged**